src/lib/stores/api-pagination.ts

```diff
@@ -1,140 +1,217 @@
 import { derived, writable } from 'svelte/store';
 import { page } from '$app/stores';
-import { perPageFromSearchParameter, defaultItemsPerPage } from './pagination';
+import {
+  perPageFromSearchParameter,
+  defaultItemsPerPage,
+  options,
+  perPageKey,
+} from './pagination';
 
 import type { Readable } from 'svelte/store';
 
 type PaginationMethods = {
+  nextPageWithItems: (t: NextPageToken, items: any[]) => void;
   nextPage: () => void;
   previousPage: () => void;
-  setNextPageToken: (t: string, items: any[]) => void;
   setUpdating: () => void;
   reset: () => void;
   resetPageSize: () => void;
+  nextRow: () => void;
+  previousRow: () => void;
+  setActiveIndex: (activeIndex: number) => void;
 };
 
 type PaginationItems = {
   key: string;
-  index: number;
-  nextIndex: number;
-  pageSize: number;
-  currentPageNumber: number;
-  endingPageNumber: number;
-  items: any[];
   loading: boolean;
   updating: boolean;
-  indexTokens: Record<number, string>;
   hasNext: boolean;
   hasPrevious: boolean;
+  index: number;
+  pageSize: number;
+  indexData: Record<
+    number,
+    { nextToken: string; start: number; end: number; items: any[] }
+  >;
+  visibleItems: any[];
+  hasNextIndexData: boolean;
+  indexStart: number;
+  indexEnd: number;
+  activeIndex: number;
 };
 
-const initialStore: PaginationItems = {
-  key: 'per-page',
-  index: 0,
-  nextIndex: 0,
-  pageSize: defaultItemsPerPage,
-  currentPageNumber: 1,
-  endingPageNumber: defaultItemsPerPage,
-  items: [],
+const defaultStore: PaginationItems = {
+  key: perPageKey,
   loading: true,
-  updating: true,
-  indexTokens: {},
-  hasNext: true,
+  updating: false,
+  hasNext: false,
   hasPrevious: false,
+  index: 0,
+  pageSize: defaultItemsPerPage,
+  indexData: {},
+  visibleItems: [],
+  hasNextIndexData: false,
+  indexStart: 0,
+  indexEnd: 0,
+  activeIndex: 0,
 };
 
 export type PaginationStore = PaginationMethods & Readable<PaginationItems>;
 
-export function createPaginationStore(): PaginationStore {
-  const paginationStore = writable(initialStore);
+export function createPaginationStore(
+  pageSizeOptions: string[] = options,
+): PaginationStore {
+  // TODO, use first option in pageSizeOptions for pageSize instead of defaultItemsPerPage
+  const initialPageSize = parseInt(pageSizeOptions?.[0]) ?? defaultItemsPerPage;
```

Comment:
What happens if `parseInt` gives us `NaN`?  I know we're all _real_ excited about `??`, but we might be getting carried away.

```js
NaN ?? 'lol' === NaN
NaN || 'lol' === 'lol'
```

But again, my general advice around pulling this logic out into smaller helper functions that we can test applies here.